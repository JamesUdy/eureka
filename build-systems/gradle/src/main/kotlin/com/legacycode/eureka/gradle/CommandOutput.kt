package com.legacycode.eureka.gradle

@JvmInline
value class CommandOutput(val content: String)
