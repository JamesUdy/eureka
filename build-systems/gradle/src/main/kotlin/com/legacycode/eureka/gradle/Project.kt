package com.legacycode.eureka.gradle

@JvmInline
value class Project(val name: String)
