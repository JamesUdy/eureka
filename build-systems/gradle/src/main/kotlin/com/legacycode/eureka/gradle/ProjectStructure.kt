package com.legacycode.eureka.gradle

data class ProjectStructure(
  val rootProject: Project,
  val subprojects: List<Project>,
) {
  companion object {
    fun from(output: CommandOutput): ProjectStructure {
      val lines = output.content.lines()
      val rootProject = parseRootProject(lines)

      val subprojects = mutableListOf<Project>()
      for (line in lines) {
        if (isProjectLine(line)) {
          subprojects.add(parseSubproject(line))
        }
      }

      return ProjectStructure(rootProject, subprojects.toList())
    }

    private fun parseRootProject(lines: List<String>): Project {
      val rootProjectLine = lines.firstOrNull { it.startsWith("Root project '") }
      val rootProjectName = rootProjectLine?.substringAfter("Root project '")?.substringBefore("'")!!
      return Project(rootProjectName)
    }

    private fun isProjectLine(line: String): Boolean {
      return line.startsWith("+--- Project") || line.startsWith("\\--- Project")
    }

    private fun parseSubproject(line: String): Project {
      val path = line.substring(line.indexOf(':') + 1).trim()
      return Project(path.substring(0, path.length - 1))
    }
  }
}
