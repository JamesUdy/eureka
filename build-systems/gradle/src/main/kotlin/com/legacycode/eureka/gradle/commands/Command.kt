package com.legacycode.eureka.gradle.commands

import java.io.File
import org.buildobjects.process.ProcBuilder

interface Command {
  val text: String

  fun execute(cmdFile: File? = null): String {
    val cmdArgs = text.split(" ")
    val cmd = cmdFile?.path ?: cmdArgs.first()
    val args = cmdArgs.drop(1).toTypedArray()

    println("Executing: $text")
    return ProcBuilder(cmd)
      .withNoTimeout()
      .withArgs(*args)
      .run()!!
      .outputString
  }
}
