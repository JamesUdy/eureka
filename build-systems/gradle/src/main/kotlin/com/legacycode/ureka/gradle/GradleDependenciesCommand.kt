package com.legacycode.ureka.gradle

class GradleDependenciesCommand(
  private val project: Project,
) {
  companion object {
    fun from(project: Project): GradleDependenciesCommand {
      return GradleDependenciesCommand(project)
    }

    fun from(projectStructure: ProjectStructure): List<GradleDependenciesCommand> {
      return projectStructure.subprojects.map(GradleDependenciesCommand::from)
    }
  }

  val text: String
    get() = "./gradlew -q ${project.name}:dependencies"
}
