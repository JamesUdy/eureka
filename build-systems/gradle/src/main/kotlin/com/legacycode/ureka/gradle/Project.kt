package com.legacycode.ureka.gradle

class Project(val name: String) {
    val subprojects: MutableList<Project> = mutableListOf()
    var parentProject: Project? = null
}
