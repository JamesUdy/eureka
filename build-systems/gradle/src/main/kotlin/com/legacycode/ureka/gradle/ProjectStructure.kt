package com.legacycode.ureka.gradle

class ProjectStructure(private val rootProject: Project) {
  companion object {
    fun fromCommandLineOutput(output: String): ProjectStructure {
      val lines = output.lines()
      val rootProjectLine = lines.firstOrNull { it.startsWith("Root project '") }
      val rootProjectName = rootProjectLine?.substringAfter("Root project '")?.substringBefore("'")
      val rootProject = Project(rootProjectName ?: "Signal")
      val projectStructure = ProjectStructure(rootProject)

      for (line in lines) {
        if (line.startsWith("+--- Project") || line.startsWith("\\--- Project")) {
          val path = line.substring(line.indexOf(':') + 1).trim()
          projectStructure.addProject(path)
        }
      }

      return projectStructure
    }
  }

  private val projects: MutableMap<String, Project> = mutableMapOf()

  fun addProject(name: String) {
    val project = Project(name.substring(0, name.length - 1))
    projects[name] = project

    val parentPath = getParentPath(name)
    if (parentPath != null) {
      val parentProject = projects[parentPath]
      parentProject?.subprojects?.add(project)
      project.parentProject = parentProject
    } else {
      rootProject.subprojects.add(project)
      project.parentProject = rootProject
    }
  }

  val printableProjectStructure: String
    get() {
      val projectStructureBuilder = StringBuilder()
      buildPrintableProjectStructure(rootProject, "", "", projectStructureBuilder)
      return projectStructureBuilder.toString()
    }

  private fun getParentPath(path: String): String? {
    val separatorIndex = path.lastIndexOf(':')
    return if (separatorIndex != -1) path.substring(0, separatorIndex) else null
  }

  private fun buildPrintableProjectStructure(
    project: Project,
    indent: String,
    branchIndent: String,
    projectStructureBuilder: StringBuilder,
  ) {
    if (project == rootProject) {
      projectStructureBuilder.appendLine("Root project '${project.name}'")
    } else {
      projectStructureBuilder.appendLine("${indent}Project ':${project.name}'")
    }
    val subProjects = project.subprojects
    for (i in subProjects.indices) {
      val subProject = subProjects[i]
      val nextBranchIndent = if (i == subProjects.lastIndex) "$branchIndent   " else "$branchIndent|  "
      if (i == subProjects.lastIndex) {
        buildPrintableProjectStructure(subProject, "$branchIndent\\--- ", nextBranchIndent, projectStructureBuilder)
      } else {
        buildPrintableProjectStructure(subProject, "$branchIndent+--- ", nextBranchIndent, projectStructureBuilder)
      }
    }
  }
}
