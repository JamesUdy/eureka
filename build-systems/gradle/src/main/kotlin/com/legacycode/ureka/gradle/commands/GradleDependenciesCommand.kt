package com.legacycode.ureka.gradle.commands

import com.legacycode.ureka.gradle.Project
import com.legacycode.ureka.gradle.ProjectStructure
import java.io.File

class GradleDependenciesCommand(
  private val root: File,
  val project: Project,
) : Command {
  companion object {
    fun from(root: File, project: Project): GradleDependenciesCommand {
      return GradleDependenciesCommand(root, project)
    }

    fun from(root: File, projectStructure: ProjectStructure): List<GradleDependenciesCommand> {
      return projectStructure.subprojects.map { from(root, it) }
    }
  }

  override val text: String
    get() = "./gradlew -p ${root.absolutePath} -q ${project.name}:dependencies"
}
