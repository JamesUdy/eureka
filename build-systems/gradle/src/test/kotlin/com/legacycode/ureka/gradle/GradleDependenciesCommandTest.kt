package com.legacycode.ureka.gradle

import com.google.common.truth.Truth.assertThat
import org.approvaltests.Approvals
import org.junit.jupiter.api.Test

class GradleDependenciesCommandTest {
  @Test
  fun `create command from subproject`() {
    // given
    val project = Project("contacts")

    // when
    val command = GradleDependenciesCommand.from(project)

    // then
    assertThat(command.text)
      .isEqualTo("./gradlew -q contacts:dependencies")
  }

  @Test
  fun `create multiple commands from project structure`() {
    // given
    val projectsTaskOutput = TaskOutputResource("projects.txt").content
    val projectStructure = ProjectStructure.from(projectsTaskOutput)

    // when
    val commands = GradleDependenciesCommand.from(projectStructure)

    // then
    val listOfCommands = commands.joinToString("\n") { it.text }
    Approvals.verify(listOfCommands)
  }
}
