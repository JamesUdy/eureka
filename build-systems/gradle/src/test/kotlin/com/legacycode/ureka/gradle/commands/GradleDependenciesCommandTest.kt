package com.legacycode.ureka.gradle.commands

import com.google.common.truth.Truth.assertThat
import com.legacycode.eureka.gradle.Project
import com.legacycode.eureka.gradle.ProjectStructure
import com.legacycode.eureka.gradle.commands.GradleDependenciesCommand
import com.legacycode.ureka.gradle.testing.TaskOutputResource
import java.io.File
import org.approvaltests.Approvals
import org.junit.jupiter.api.Test

class GradleDependenciesCommandTest {
  @Test
  fun `create command from subproject`() {
    // given
    val project = Project("contacts")

    // when
    val command = GradleDependenciesCommand.from(File("/project/root"), project)

    // then
    assertThat(command.text)
      .isEqualTo("./gradlew -p /project/root -q contacts:dependencies")
  }

  @Test
  fun `create multiple commands from project structure`() {
    // given
    val projectsTaskOutput = TaskOutputResource.projects("Signal.txt").output
    val projectStructure = ProjectStructure.from(projectsTaskOutput)

    // when
    val commands = GradleDependenciesCommand.from(File("/project/root"), projectStructure)

    // then
    val listOfCommands = commands.joinToString("\n") { it.text }
    Approvals.verify(listOfCommands)
  }
}
