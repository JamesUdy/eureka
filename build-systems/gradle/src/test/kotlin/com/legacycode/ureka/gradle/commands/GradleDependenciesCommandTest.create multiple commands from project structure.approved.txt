./gradlew -p /project/root -q benchmark:dependencies
./gradlew -p /project/root -q contacts:dependencies
./gradlew -p /project/root -q contacts-app:dependencies
./gradlew -p /project/root -q core-ui:dependencies
./gradlew -p /project/root -q core-util:dependencies
./gradlew -p /project/root -q device-transfer:dependencies
./gradlew -p /project/root -q device-transfer-app:dependencies
./gradlew -p /project/root -q donations:dependencies
./gradlew -p /project/root -q donations-app:dependencies
./gradlew -p /project/root -q glide-config:dependencies
./gradlew -p /project/root -q image-editor:dependencies
./gradlew -p /project/root -q image-editor-app:dependencies
./gradlew -p /project/root -q libsignal-service:dependencies
./gradlew -p /project/root -q lintchecks:dependencies
./gradlew -p /project/root -q microbenchmark:dependencies
./gradlew -p /project/root -q paging:dependencies
./gradlew -p /project/root -q paging-app:dependencies
./gradlew -p /project/root -q photoview:dependencies
./gradlew -p /project/root -q qr:dependencies
./gradlew -p /project/root -q qr-app:dependencies
./gradlew -p /project/root -q Signal-Android:dependencies
./gradlew -p /project/root -q sms-exporter:dependencies
./gradlew -p /project/root -q sms-exporter-app:dependencies
./gradlew -p /project/root -q spinner:dependencies
./gradlew -p /project/root -q spinner-app:dependencies
./gradlew -p /project/root -q sticky-header-grid:dependencies
./gradlew -p /project/root -q video:dependencies
